server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -1778,4 +1740,124 @@ List<SearchShardIterator> getLocalShardsIterator(
+                    ShardOperationFailedException[] groupedFailures = ExceptionsHelper.groupBy(searchResponse.getShardFailures());
+                    for (ShardOperationFailedException f : groupedFailures) {
+                        boolean causeHas500Status = false;
+                        if (f.getCause() != null) {
+                            causeHas500Status = ExceptionsHelper.status(f.getCause()).getStatus() >= 500;
+                        }
+                        if ((f.status().getStatus() >= 500 || causeHas500Status)
+                            && ExceptionsHelper.isNodeOrShardUnavailableTypeException(f.getCause()) == false) {
+                            logger.warn("TransportSearchAction shard failure (partial results response)", f);
+                            responseCountTotalStatus = SearchResponseMetrics.ResponseCountTotalStatus.PARTIAL_FAILURE;
+                        }
+                    }
+                }
+                // increment after the delegated onResponse to ensure we don't
+                // record both a success and a failure if there is an exception
+                searchResponseMetrics.incrementResponseCount(responseCountTotalStatus);
+
+                if (collectTelemetry()) {
+                    extractCCSTelemetry(searchResponse);
+                    recordTelemetry();
+                }
+                // TODO: should this be last?
+                listener.onResponse(searchResponse);
+            } catch (Exception e) {
+                onFailure(e);
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            searchResponseMetrics.incrementResponseCount(SearchResponseMetrics.ResponseCountTotalStatus.FAILURE);
+            if (collectTelemetry()) {
+                // TODO: better failure recognition
+                Result status;
+                if (e instanceof RemoteTransportException) {
+                    status = Result.REMOTES_UNAVAILABLE;
+                } else if (task.isCancelled() || (e instanceof TaskCancelledException)) {
+                    status = Result.CANCELED;
+                } else {
+                    status = Result.UNKNOWN;
+                }
+                usageBuilder.setFailure(status);
+                // TODO: can we still get some time measurements here? Do we want to?
+                recordTelemetry();
+            }
+            listener.onFailure(e);
+        }
+
+        private void recordTelemetry() {
+            usageService.getCcsUsageHolder().updateUsage(usageBuilder.build());
+        }
+
+        /**
+         * Extract telemetry data from the search response.
+         * @param searchResponse The final response from the search.
+         */
+        private void extractCCSTelemetry(SearchResponse searchResponse) {
+            usageBuilder.took(searchResponse.getTookInMillis());
+            // TODO: what happens with the local cluster there? Are we tracking it too just like the others?
```

Comment:
Good question - I don't think we need/want to track the local cluster individually, but probably best to check with Product.

Reply:
For now I am assuming it would just produce `(local)` entry and I leave it as is, but I'll check it later and it would be easy to add a check here that removes the local one.